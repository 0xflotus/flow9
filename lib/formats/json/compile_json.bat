del json_pegop.flow
call flowcpp lingo/pegcode/pegcompiler.flow -- "file=../../../lib/formats/json/json.lingo" out=json_pegop.flow
pause