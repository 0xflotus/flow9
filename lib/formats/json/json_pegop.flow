// Generated by
//   flowcpp lingo/pegcode/pegcompiler.flow -- file=../../../lib/formats/json/json.lingo out=json_pegop.flow

import lingo/pegcode/pegcode;

export {
	pegOps4Json : [PegOp];
}

pegOps4Json : [PegOp] =
	[PegCall(2), PegEnd(), PegCall(222), PegDrop(), PegCall(47), PegReturn(),
	PegString("{"), PegDrop(), PegCall(222), PegDrop(), PegStartCaptureList(), PegChoice(3), PegCall(23), PegCommit(1), PegCaptureList(), PegBindValue("v0"), PegDrop(), PegString("}"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("JsonObject", [ActionCall("unarray", [ActionVar("v0")])])), PegReturn(),
	PegCall(35), PegBindValue("v0"), PegDrop(), PegStartCaptureList(), PegChoice(3), PegCall(233), PegCommit(-2), PegCaptureList(), PegBindValue("v1"), PegDrop(), PegAction2(ActionCallDebug("consArray", [ActionVar("v0"), ActionVar("v1")])), PegReturn(),
	PegCall(204), PegBindValue("v0"), PegDrop(), PegString(":"), PegDrop(), PegCall(222), PegDrop(), PegCall(47), PegBindValue("v1"), PegDrop(), PegAction2(ActionCallDebug("Pair", [ActionVar("v0"), ActionVar("v1")])), PegReturn(),
	PegChoice(3), PegCall(6), PegCommit(29), PegChoice(3), PegCall(106), PegCommit(26), PegChoice(6), PegCall(204), PegBindValue("v0"), PegDrop(), PegAction2(ActionCallDebug("JsonString", [ActionVar("v0")])), PegCommit(20), PegChoice(8), PegCall(119), PegBindValue("v0"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("JsonDouble", [ActionVar("v0")])), PegCommit(12), PegChoice(6), PegCall(79), PegBindValue("v0"), PegDrop(), PegAction2(ActionCallDebug("JsonArray", [ActionVar("v0")])), PegCommit(6), PegString("null"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("JsonNull", [])), PegReturn(),
	PegString("["), PegDrop(), PegCall(222), PegDrop(), PegChoice(17), PegCall(47), PegBindValue("v0"), PegDrop(), PegStartCaptureList(), PegChoice(3), PegCall(243), PegCommit(-2), PegCaptureList(), PegBindValue("v1"), PegDrop(), PegString("]"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("consArray", [ActionVar("v0"), ActionVar("v1")])), PegCommit(6), PegString("]"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("array", [])), PegReturn(),
	PegChoice(7), PegString("true"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("JsonBool", [ActionCall("true", [])])), PegCommit(6), PegString("false"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("JsonBool", [ActionCall("false", [])])), PegReturn(),
	PegChoice(7), PegStartString(), PegCall(253), PegBindString("v0"), PegDrop(), PegAction2(ActionCallDebug("s2d", [ActionVar("v0")])), PegCommit(13), PegChoice(7), PegStartString(), PegCall(263), PegBindString("v0"), PegDrop(), PegAction2(ActionCallDebug("s2d", [ActionVar("v0")])), PegCommit(6), PegStartString(), PegCacheCall(146, 9), PegBindString("v0"), PegDrop(), PegAction2(ActionCallDebug("s2d", [ActionVar("v0")])), PegReturn(),
	PegStartCaptureList(), PegCall(153), PegChoice(3), PegCall(153), PegCommit(-2), PegCaptureList(), PegReturn(),
	PegChoice(5), PegString("-"), PegDrop(), PegCall(139), PegCommit(2), PegCall(139), PegReturn(),
	PegRange("0", "9"), PegReturn(),
	PegChoice(3), PegCall(153), PegCommit(3), PegRangeChoice("a", "f", 2), PegRange("A", "F"), PegReturn(),
	PegChoice(11), PegString("\\u"), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegCommit(30), PegChoice(11), PegString("\\X"), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegCommit(19), PegChoice(7), PegString("\\x"), PegDrop(), PegCall(155), PegDrop(), PegCall(155), PegCommit(12), PegChoice(5), PegString("\\"), PegDrop(), PegCall(202), PegCommit(7), PegChoice(4), PegString("\""), PegCommit(1), PegFail(), PegDrop(), PegCall(202), PegReturn(),
	PegRange("\u0000", "\uffff"), PegReturn(),
	PegString("\""), PegDrop(), PegStartString(), PegStartCaptureList(), PegChoice(3), PegCall(161), PegCommit(-2), PegCaptureList(), PegBindString("v0"), PegDrop(), PegStartString(), PegString("\""), PegBindString("v1"), PegDrop(), PegCall(222), PegDrop(), PegAction2(ActionCallDebug("strUnescape", [ActionVar("v0"), ActionVar("v1")])), PegReturn(),
	PegStartCaptureList(), PegChoice(3), PegCall(228), PegCommit(-2), PegCaptureList(), PegReturn(),
	PegStringChoice(" ", 4), PegStringChoice("\t", 3), PegStringChoice("\n", 2), PegString("\u000d"), PegReturn(),
	PegString(","), PegDrop(), PegCall(222), PegDrop(), PegCall(35), PegClearCache(), PegBindValue("v0"), PegDrop(), PegAction2(ActionVar("v0")), PegReturn(),
	PegString(","), PegDrop(), PegCall(222), PegDrop(), PegCall(47), PegClearCache(), PegBindValue("v0"), PegDrop(), PegAction2(ActionVar("v0")), PegReturn(),
	PegCacheCall(146, 9), PegDrop(), PegString("."), PegDrop(), PegStartCaptureList(), PegChoice(3), PegCall(139), PegCommit(1), PegCaptureList(), PegReturn(),
	PegString("."), PegDrop(), PegCall(139), PegReturn()];
