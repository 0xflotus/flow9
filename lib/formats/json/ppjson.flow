import formats/json/json;
import pretty;

export {
	// It is probably better to use json2stringFormatted instead of this. 
	json2doc(j : Json) -> Doc;
}

json2doc(j : Json) -> Doc {
	switch(j) {
		JsonNull(): textDoc("null");
		JsonBool(v): textDoc(b2s(v));
		JsonDouble(d): textDoc(d2s(d));
		JsonString(s): textDoc(toString(s));
		JsonArray(arr): {
			catDocs([
				textDoc("["),
				nestDoc(1, lineDoc()),
				nestDoc(1,
					catDocs(interleave(
						map(arr, json2doc),
						catDoc(textDoc(","), lineDoc())
					))
				),
				lineDoc(),
				textDoc("]")
			])
		}
		JsonObject(members): {
			numberOfMembers = length(members);
			catDocs([
				textDoc("{"),
				nestDoc(1, lineDoc()),
				nestDoc(1,
					catDocs(interleave(
						mapi(members, \idx, m -> {
							catDocs([
								textDoc(toString(m.first)),
								textDoc(": "),
								json2doc(m.second),
								if (idx < numberOfMembers - 1) textDoc(",") else NilDoc()
							])
						}),
						lineDoc()
					))
				),
				lineDoc(),
				textDoc("}"),
			])
		}
	}
}