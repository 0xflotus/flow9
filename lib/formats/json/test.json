{
    "key": {
    	"object": [1, 2, 3, {
    		"try": "string"
    	}]
	}, 
	"k2": {},
	"k3": [[],-1.0],
	"k3": "tab\ttab",
	"k4": "quote\"quote"
}