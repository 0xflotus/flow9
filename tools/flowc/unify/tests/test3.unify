import tuples;
import json;

α24 = Pair<string, JsonObject>;
α11 = .(.second : α29, .first : α25);
α29 = JsonObject<>;
α24  c=  α11;
