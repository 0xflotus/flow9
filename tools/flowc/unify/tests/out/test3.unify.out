α11=Pair<string, JsonObject<>>
α24=Pair<string, JsonObject<>>
α25=string
α29=JsonObject<>
